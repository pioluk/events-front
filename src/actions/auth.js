/* @flow */

import {
  LOGIN_REQUEST,
  LOGIN_SUCCESS,
  LOGIN_FAILURE,
  LOGOUT
} from '../constants/auth'

export function loginRequest (username: string, password: string) {
  return {
    type: LOGIN_REQUEST,
    username,
    password
  }
}

export function loginSuccess (user: string, token: string) {
  return {
    type: LOGIN_SUCCESS,
    user,
    token
  }
}

export function loginFailure (error: Error) {
  return {
    type: LOGIN_FAILURE,
    error
  }
}

export function logout () {
  return {
    type: LOGOUT
  }
}
