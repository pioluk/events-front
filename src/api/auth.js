/* @flow */

import { apiPost } from './utils'

export const login = (username: string, password: string) =>
  apiPost('login', { username, password })
