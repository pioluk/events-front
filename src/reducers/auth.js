/* @flow */

import {
  LOGIN_REQUEST,
  LOGIN_SUCCESS,
  LOGIN_FAILURE,
  LOGOUT
} from '../constants/auth'

type AuthState = {
  errorMessage: ?string,
  isAuthenticated: boolean,
  isAuthenticating: boolean,
  token: ?string,
  user: ?any
}

const initialState: AuthState = {
  errorMessage: null,
  isAuthenticated: false,
  isAuthenticating: false,
  token: null,
  user: null
}

export default function authReducer (state: AuthState = initialState, action: any) {
  switch (action.type) {
    case LOGIN_REQUEST:
      return Object.assign({}, state, { isAuthenticating: true })

    default:
      return state
  }
}
