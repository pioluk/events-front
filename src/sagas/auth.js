import { call, take, put } from 'redux-saga/effects'
import { browserHistory } from 'react-router'
import { LOGIN_REQUEST } from '../constants/auth'
import * as authActions from '../actions/auth'
import * as notificationActions from '../actions/notifications'
import { login } from '../api/auth'

export function* loginFlow () {
  while (true) {
    const { username, password } = yield take(LOGIN_REQUEST);
    try {
      const { user, token } = yield call(login, username, password);
      if (user && token) {
        localStorage.setItem('user', JSON.stringify(user));
        localStorage.setItem('token', token);
        yield put(authActions.loginSuccess(user, token));
        yield put(notificationActions.displayNotification('Successfully logged in.'))
        browserHistory.push('/');
      }
    }
    catch (err) {
      yield put(authActions.loginFailure(err))
      yield put(notificationActions.displayNotification('Error: ' + err.message))
    }
  }
}
