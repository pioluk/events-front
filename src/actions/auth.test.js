// global describe,test,expect

import * as actions from './auth'

describe('auth actions', () => {
  test('it should create an action to send a login request', () => {
    const username = 'admin'
    const password = 'admin1'
    expect(actions.loginRequest(username, password)).toMatchSnapshot()
  })

  test('it should create an action to indicate login success', () => {
    const user = 'admin'
    const token = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiYWRtaW4iOnRydWV9.TJVA95OrM7E2cBab30RMHrHDcEfxjoYZgeFONFh7HgQ'
    expect(actions.loginSuccess(user, token)).toMatchSnapshot()
  })

  test('it should create an action to indicate login failure', () => {
    const error = new Error('Login failure')
    expect(actions.loginFailure(error)).toMatchSnapshot()
  })

  test('it should create an action to logout', () => {
    expect(actions.logout()).toMatchSnapshot()
  })
})
