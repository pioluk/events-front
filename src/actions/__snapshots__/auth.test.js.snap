exports[`auth actions it should create an action to indicate login failure 1`] = `
Object {
  "error": [Error: Login failure],
  "type": "LOGIN_FAILURE",
}
`;

exports[`auth actions it should create an action to indicate login success 1`] = `
Object {
  "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiYWRtaW4iOnRydWV9.TJVA95OrM7E2cBab30RMHrHDcEfxjoYZgeFONFh7HgQ",
  "type": "LOGIN_SUCCESS",
  "user": "admin",
}
`;

exports[`auth actions it should create an action to logout 1`] = `
Object {
  "type": "LOGOUT",
}
`;

exports[`auth actions it should create an action to send a login request 1`] = `
Object {
  "password": "admin1",
  "type": "LOGIN_REQUEST",
  "username": "admin",
}
`;
